netbsd/hid.c

```diff
@@ -0,0 +1,1121 @@
+	int usage_found = 0;
+
+	while (*pos < size) {
+		int key = report_descriptor[*pos];
+		int key_cmd = key & 0xfc;
+
+		/* Determine data_len and key_size */
+		if (!get_hid_item_size(report_descriptor, *pos, size, &data_len, &key_size))
+			return -1; /* malformed report */
+
+		switch (key_cmd) {
+		case 0x4: /* Usage Page 6.2.2.7 (Global) */
+			*usage_page = get_hid_report_bytes(report_descriptor, size, data_len, *pos);
+			break;
+
+		case 0x8: /* Usage 6.2.2.8 (Local) */
+			*usage = get_hid_report_bytes(report_descriptor, size, data_len, *pos);
+			usage_found = 1;
+			break;
+
+		case 0xa0: /* Collection 6.2.2.4 (Main) */
+			/* A Usage Item (Local) must be found for the pair to be valid */
+			if (usage_found)
+				usage_pair_ready = 1;
+
+			/* Usage is a Local Item, unset it */
+			usage_found = 0;
+			break;
+
+		case 0x80: /* Input 6.2.2.4 (Main) */
+		case 0x90: /* Output 6.2.2.4 (Main) */
+		case 0xb0: /* Feature 6.2.2.4 (Main) */
+		case 0xc0: /* End Collection 6.2.2.4 (Main) */
+			/* Usage is a Local Item, unset it */
+			usage_found = 0;
+			break;
+		}
+
+		/* Skip over this key and its associated data */
+		*pos += data_len + key_size;
+
+		/* Return usage pair */
+		if (usage_pair_ready)
+			return 0;
+	}
+
+	/* If no top-level application collection is found and usage page/usage pair is found, pair is valid
+	   https://docs.microsoft.com/en-us/windows-hardware/drivers/hid/top-level-collections */
+	if (initial && usage_found)
+		return 0; /* success */
+
+	return 1; /* finished processing */
+}
+
+static struct hid_device_info *create_device_info(const struct usb_device_info *udi, const char *path, const struct usb_ctl_report_desc *ucrd)
+{
+	struct hid_device_info *root;
+	struct hid_device_info *end;
+
+	root = (struct hid_device_info *) calloc(1, sizeof(struct hid_device_info));
+	if (!root)
+		return NULL;
+
+	end = root;
+
+	/* Path */
+	end->path = (path) ? strdup(path) : NULL;
+
+	/* Vendor Id */
+	end->vendor_id = udi->udi_vendorNo;
+
+	/* Product Id */
+	end->product_id = udi->udi_productNo;
+
+	/* Serial Number */
+	end->serial_number = utf8_to_wchar_t(udi->udi_serial);
+
+	/* Release Number */
+	end->release_number = udi->udi_releaseNo;
+
+	/* Manufacturer String */
+	end->manufacturer_string = utf8_to_wchar_t(udi->udi_vendor);
+
+	/* Product String */
+	end->product_string = utf8_to_wchar_t(udi->udi_product);
+
+	/* Usage Page */
+	end->usage_page = 0;
+
+	/* Usage */
+	end->usage = 0;
+
+	/* Interface Number */
+	end->interface_number = -1;
+
+	/* Next Device Info */
+	end->next = NULL;
+
+	/* Bus Type */
+	end->bus_type = HID_API_BUS_USB;
+
+	if (ucrd) {
+		uint32_t pos;
+		uint16_t page;
+		uint16_t usage;
+
+		pos = page = usage = 0;
+
+		/*
+		 * Parse the first usage and usage page
+		 * out of the report descriptor.
+		 */
+		if (!get_next_hid_usage(ucrd->ucrd_data, ucrd->ucrd_size, &pos, &page, &usage)) {
```

Comment:
this is probably a copy-paste (or not), but since technically we're writing this from scratch - I'd prefer reading such conditionas as:
```suggestion
		if (get_next_hid_usage(ucrd->ucrd_data, ucrd->ucrd_size, &pos, &page, &usage) == 0) {
```
generally speaking, having a negation (`!`) automatically triggers human brain to look for a failed call (even if often not the case in unix world)